features/123-chat-push-notification/api.md

```diff
@@ -1,158 +1,108 @@
-  - `skygearChat.typingStarted((conversation, participants, events) => {})`
-  - `skygearChat.afterConversationCreated((conversation, participants) => {})`
-  - `skygearChat.afterConversationUpdated((conversation, participants) => {})`
-  - `skygearChat.afterConversationDeleted((conversation, participants) => {})`
-  - `skygearChat.afterUsersAddedToConversation((conversation, participants, users) => {})`
-  - `skygearChat.afterUsersRemovedFromConversation((conversation, participants, users) => {})`
-
-- Conversation
-  - `@chat.after_message_sent`
-  - `@chat.after_message_updated`
-  - `@chat.after_message_deleted`
-  - `@chat.typing_started`
-  - `@chat.after_conversation_created`
-  - `@chat.after_conversation_updated`
-  - `@chat.after_conversation_deleted`
-  - `@chat.after_users_added_to_conversation`
-  - `@chat.after_users_removed_from_conversation`
+## Task
 
-## Implementation Details
+1. Define new payload format
+2. Update skygear server to support new payload.
 
+## New Payload
 
-### Changes on JS SDK
-- Example:
+The following is the proposed payload.
 
-```javascript
-export function afterMessageSent(callback) {
-  registry.registerHandler('chat:after_message_sent', (req) => {
-    var conversation = blahblahblah ;
-    var message = blahblahblah;
-    callback(message, conversation);
-  }, {});
-}
-```
+### Main Payload
 
-### Changes on Plugin (Hook Wrapper)
-- Wrapper of op wrapper
+Main Payload is a collection of payloads.
 
-- Example:
+|key | type |description           |default|
+|----|------|----------------------|-------|
+|platform|dict|See platform payload|null|
+|message|dict|See message payload|null|
+|data|dict|Custom dictionary by developer|null|
+|custom|dict|See custom payload|null|
+|dry\_run|boolean|dry run only if true|false|
 
-```python
-def after_message_sent(*args, **kwargs):
-    def chat_op(func):
-        _registry.register_op('chat:after_message_sent', func, *args, **kwargs)
-        return func
-    return chat_op
-```
 
 
-### Changes on Plugin (Database Operation)
-- Send message to call specific hook after database operation
-- Example:
+### Platform Payload
 
-``` python
-container.send_action('chat:after_conversation_hook', {'roles': message, 'conversation': conversation} )
-```
+|key | type |description|default|
+|----|------|-----------|-------|
+|android|boolean|send to android device if true|true|
+|ios|boolean|send to android device if true|true|
+
+### Message Payload
+
+|key|type|description|default|
+|---|----|-----------|-------|
+|body|string|message body|null|
+|title|string|message title|null|
+|subtitle|string|message subtitle|null|
+|sound|string|message sound|null|
+|title\_loc\_key|string|key to localized string|null|
+|title\_loc\_args|list of string|arguments of string replacement of title\_loc\_key|null|
+|priority|string|priority of message|"normal"|
+|expiry|int|time to live in seconds| 2419200 (4 weeks)|
```

Comment:
I would suggest to make it a determine datetime. It will make debugging easier and determine.

Reply:
I think keeping it as seconds make more sense (and easier to trace) as in FCM & APNS it is also in number of seconds.
https://firebase.google.com/docs/cloud-messaging/concept-options

Reply:
In the link you provide, there is ttl and expiry.

If the name is ttl, I am fine with the second.
if the name is expiry, better to make it Epoch Unix time stamp.

Reply:
ok